### ISSUE TITLE
MicroDrop UI freezes when running protocol with pump enabled

### ISSUE THREAD

**Issue by cfobel** [wheeler-microfluidics/microdrop.mr-box-plugin#4]
# Set up
 
 1. Enable `mr_box_plugin` and `dmf_device_ui_plugin`.
 2. Run protocol with pump enabled on at least one step.
 3. Close pump prompt after pumping is complete.

Here is an example protocol:

![image](https://user-images.githubusercontent.com/823355/29138159-c13c1de6-7d10-11e7-885c-e189439966c8.png)

Download the above protocol here: [freeze test (christian).zip](https://github.com/wheeler-microfluidics/microdrop.mr-box-plugin/files/1212501/freeze.test.christian.zip)

# Expected behaviour

 - Pump dialog closes.
 - Protocol continues to execute.

# Observed behaviour

User interface stops responding and protocol does not continue.

**Comment by cfobel** [comment]
# Cause

This behaviour is caused by two main problems:

 1. MicroDrop and some other plugins not handling threading properly.
 2. Plugins calling `hub_execute` and running `refresh_gui` as a wait function.

Regarding **(1)**, **all** GTK calls **MUST** be run in the **main GTK thread**.  This can be accomplished, e.g., by using the [`gobject.idle_add(...)`][1] function to schedule a function to be called from the main GTK thread.

See [here][1] for more information regarding threading and pygtk.

Regarding **(2)**, as of [`microdrop` `v2.10`][2], calls to `hub_execute` would not complete until the GTK main loop was able to iterate.  However, since plugins execute within the GTK main loop **and** the `device_info_plugin` ZeroMQ sockets which are used to process all `hub_execute` commands are monitored in the GTK main loop through `gobject.timeout_add` calls, plugins needed to explicitly use the `pygtkhelpers.utils.refresh_gui` to allow the GTK thread to process pending events.  Unfortunately, this leads to freezing the GTK loop.

[1]: http://faq.pygtk.org/index.py?req=show&file=faq20.001.htp
[2]: https://github.com/wheeler-microfluidics/microdrop/releases/tag/v2.10

# Fix

## `microdrop` ##

Modified code: [microdrop.zip](https://github.com/wheeler-microfluidics/microdrop.mr-box-plugin/files/1212660/microdrop.zip)

 - Modify device info core plugin to run a separate thread for monitoring ZeroMQ sockets (i.e., do not monitor sockets in GTK main thread).
 - Modify protocol controller to emit signals through calls to `gobject.idle_add` to ensure the signal handlers (which may interact with GTK UI objects) are run in the GTK main thread.

# DMF device UI

Modified code: [dmf_device_ui_plugin.zip](https://github.com/wheeler-microfluidics/microdrop.mr-box-plugin/files/1212658/dmf_device_ui_plugin.zip)

 - Use `gobject.idle_add` to schedule execution of methods/functions that interact with GTK UI objectsto ensure they are run in the GTK main thread.

**Comment by cfobel** [comment]
**TODO** Unrelated to this issue, but a reminder for @cfobel to update the mr-box-plugin to set the units on the ADC `pandas.Series` to show the measurement units.

i.e.: 
```python
        while True:
            data_i = MAX11210_read(proxy, rate=adc_rate,
                                   duration_s=delta_t.total_seconds())
            #Convert data to Voltage, 24bit ADC with Vref = 3.0 V
            data_i /=  ((2 ** 24 - 1)/(3.0/adc_dgain))
            #Convert Voltage to Current, 30kOhm Resistor
            data_i /= 30e3
            # Set name to display units.
            data_i.name = 'Current (A)'
            data.append(data_i)
            data_ready.set()
            if stop_event.is_set():
                break
```

**Comment by cfobel** [comment]
Fixed by wheeler-microfluidics/microdrop@b4ceb23d5d2b09f66c786d1f6852b5b588c7cf97 and wheeler-microfluidics/dmf_device_ui_plugin@3fa81d42eacceaafe22e10e3cd9715c333c43cb9.
